flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/SemanticXidGenerator.java

```diff
@@ -46,20 +50,29 @@
 
     private static final int FORMAT_ID = 201;
 
-    private transient byte[] gtridBuffer; // globalTransactionId = checkpoint id (long)
-    private transient byte[] bqualBuffer; // branchQualifier = task index + random bytes
+    private transient byte[]
+            gtridBuffer; // globalTransactionId = job id + task index + checkpoint id
+    private transient byte[] bqualBuffer; // branchQualifier = random bytes
 
     @Override
     public void open() {
-        bqualBuffer = getRandomBytes(Long.BYTES);
-        gtridBuffer = new byte[Long.BYTES];
+        gtridBuffer = new byte[3 * Long.BYTES];
+        bqualBuffer = getRandomBytes(Integer.BYTES);
     }
 
     @Override
     public Xid generateXid(RuntimeContext runtimeContext, long checkpointId) {
-        writeNumber(runtimeContext.getIndexOfThisSubtask(), gtridBuffer, 0);
-        // deliberately write only 4 bytes of checkpoint id and rely on random generation
-        writeNumber((int) checkpointId, gtridBuffer, Integer.BYTES);
+        Optional<JobID> jobId = runtimeContext.getJobId();
+        if (jobId.isPresent()) {
+            System.arraycopy(jobId.get().getBytes(), 0, gtridBuffer, 0, 16);
```

Comment:
nit: how about using public `JobID.SIZE` instead of 16 here and in other places (in declaration we can add 8 bytes)?